### ISSUE TITLE
Libdoc performance degradation starting from RF 6.0

### ISSUE THREAD

**Issue by picope** [robotframework/robotframework#5254]
In the project I'm working on we use dynamic python libraries with 10k+ keywords. After migrating to RF 6.0 we noticed that the libdoc execution time doubled. 

Here is a comparison of the time taken by libdoc in each version given the same parameters:

![image](https://github.com/user-attachments/assets/bc0da080-730f-4629-9c86-911282bc9420)

Digging into the code a bit I noticed that there is a class called `Languages` that is instantiated once for each Keyword, but could be instanciated once and use n times. Here is a pyinstrument screenshot:

![image](https://github.com/user-attachments/assets/afde0b98-6077-429e-aa47-9fcc726080c8)

## Proposed Fixes

* Suggestion 1: A pre-created instance of `Languages` could be provided when calling `TypeConverter.converter_for`:
https://github.com/robotframework/robotframework/blob/c9c9d414d8aa66959d3df7ef1895516c1d37d6c2/src/robot/libdocpkg/datatypes.py#L55

* Suggestion 2: Convert the class `Languages` into a singleton.

**Comment by pekkaklarck** [comment]
Interesting. I can reproduce the problem that recreating `Languages()` slows down Libdoc execution. In my case creating documentation for a very simple dynamic library with 10000 keywords dropped from 1.6s to 0.8s with just the following change:

```diff
--- a/src/robot/running/arguments/typeconverters.py
+++ b/src/robot/running/arguments/typeconverters.py
@@ -51,7 +51,7 @@ class TypeConverter:
                  languages: 'Languages|None' = None):
         self.type_info = type_info
         self.custom_converters = custom_converters
-        self.languages = languages or Languages()
+        self.languages = languages
```

I believe this wouldn't actually even break anything, because `TypeConverter.languages` is only used for Boolean conversion and during execution `TypeConverter` should always get a `Languages` instance instead of `None`. It would nevertheless be somewhat risky, but making `TypeConverter.languages` a property that creates the instance only when accessed ought to totally safe and give the exact same performance benefits.

Have you tested is this the only cause for performance regressions or is there something else as well?

**Comment by pekkaklarck** [comment]
I also tested by simple library with earlier Robot versions. I didn't see as big difference performance degradation as you. More importantly, with the fix the performance was much better than with earlier versions. Times below are rough averages from three executions.

| version | time  |
| ------- | ----  |
| v4.1    | ~1.4s |
| v5.0    | ~1.5s |
| v6.1    | ~2.2s |
| v7.1    | ~1.7s |
| master w/o fix | ~1.6s |
| master w/ fix  | ~0.9s |

**Comment by pekkaklarck** [comment]
If you test with the current master, notice that the underlying tech used for HTML outputs has changed totally (#4304). It shouldn't affect the performance of generating HTML outputs, but based on the above rough numbers the master was a slightly faster than v7.1 already before the fix.

**Comment by pekkaklarck** [comment]
I got the fix to this already committed locally and will push it shortly. I also decided to look why instantiating `Languages` is so slow that it has such a big performance effect. It turned out there were two main reasons:

- The code always queries what languages would be available. This is unnecessary if no custom language has been used i.e. when `Languages` is initialized without arguments. This is easy to avoid and enhances performance of calling `Language()` with close to 50%. I'll commit and push this change soon as well.
- It just takes some time to update all headers, settings, etc. when going through a single language. This is done with English also when `Languages()` is initialized without arguments. It's not easy to do anything to this. `Languages` shouldn't be initialized more than once during normal execution, so enhancements wouldn't really be worth the effort.

I tested the performance with the aforementioned performance fix and without the change to `TypeConverter` and generation time was ~1.3s. Already that would have given a nice performance boost, but the other fix is clearly much better and after that this latter fix doesn't anymore save anything. It's anyway good to know why initializing `Languages` is quite slow and that in a common case it's now quite a bit faster.

**Comment by picope** [comment]
I'm glad you were able to reproduce the case. I haven't tested whether this is the only cause for performance regressions.

Is there any chance that the fix could be introduced as a patch for version 6? maybe 6.1.2?

**Comment by pekkaklarck** [comment]
We only do bug fix releases for old versions if there are critical bugs or other special cases. This isn't an important enough issue to warrant that.

**Comment by pekkaklarck** [comment]
If the fix was to backported to RF 6.1.x, it would also need to be backported to RF 7.0.x and RF 7.1.x which would increase the effort. Backporting only to RF 7.1.x could be considered, but we don't have plans for further releases in that series.
